[VPlan] Verify dominance for incoming values of phi-like recipes.

fhahn opened the pull request:
Update the verifier to verify dominance for incoming values for phi-like recipes. The defining recipe must dominate the incoming block for the incoming value.

Builds on top of https://github.com/llvm/llvm-project/pull/138472 to retrieve incoming values & corresponding blocks for phi-like recipes.

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -175,6 +175,11 @@ bool VPlanVerifier::verifyEVLRecipe(const VPInstruction &EVL) const {
+static bool isVPIRInstructionPhi(const VPRecipeBase &R) {
+  auto *VPIRI = dyn_cast<VPIRInstruction>(&R);
+  return VPIRI && isa<PHINode>(VPIRI->getInstruction());
+}
```
nit: Add new line after function?

david-arm commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -207,22 +212,65 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+                   "Unexpected recipe with 2 or more predecessors");
+            Incoming = UserVPBB->getPredecessors()[1];
+          }
+          if (auto *R = dyn_cast<VPRegionBlock>(Incoming))
```
Just out of curiosity, what's the scenario where the incoming block for a VPWidenPHIRecipe, VPHeaderPHIRecipe or VPPredInstPHIRecipe recipe would be the region? Is it the recipes we create in the middle block? If so, could this check potentially fail with uncountable early exits, i.e. the incoming block is effectively the middle.split block? I assume it's fine, but just wondering if there are missing tests for all the scenarios.

david-arm reviewed (commented):
Thanks for this - looks sensible to me. I just had a couple of comments.

fhahn reviewed (commented):
ping :)

ayalz reviewed (commented):
Good to see this TODO being addressed, adding various comments.

fhahn committed:
[VPlan] Add VPPhiAccessors to provide interface for phi recipes (NFC)

Add a VPPhiAccessors class to provide interfaces to access incoming
values and blocks, with corresponding iterators.

The first user is VPWidenPhiRecipe, with the other phi-like recipes
following soon.

This will also be used to verify def-use chains where users are phi-like
recipes, simplifying https://github.com/llvm/llvm-project/pull/124838.

fhahn reviewed (commented):
I reworked the PR quite a bit and it is now building on a few additional patches:

* Add interface to retrieve incoming blocks & values + iterators, usable for all phi-like recipes: https://github.com/llvm/llvm-project/pull/129387
* Add VPIRPhi https://github.com/llvm/llvm-project/pull/129387

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -207,24 +213,69 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+        // Verify incoming value of various phi-like recipes.
+        if (isa<VPWidenPHIRecipe, VPHeaderPHIRecipe, VPPredInstPHIRecipe>(UI)) {
+          const VPBlockBase *Incoming = nullptr;
+          // Get the incoming block based on the number of predecessors.
```
Code has been replaced by type-switch and using incoming values/blocks iterator.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -207,24 +213,69 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+        if (isa<VPWidenPHIRecipe, VPHeaderPHIRecipe, VPPredInstPHIRecipe>(UI)) {
+          const VPBlockBase *Incoming = nullptr;
+          // Get the incoming block based on the number of predecessors.
+          if (UserVPBB->getNumPredecessors() == 0) {
```
Code is gone in latest version

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -207,24 +213,69 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+          if (UserVPBB->getNumPredecessors() == 0) {
+            assert((isa<VPWidenPHIRecipe>(UI) || isa<VPHeaderPHIRecipe>(UI)) &&
+                   "Unexpected recipe with 0 predecessors");
+            const VPRegionBlock *UserRegion = UserVPBB->getParent();
```
Code is gone in latest version

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -207,24 +213,69 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+                   "Unexpected recipe with 2 or more predecessors");
+            Incoming = UserVPBB->getPredecessors()[1];
+          }
+          if (auto *R = dyn_cast<VPRegionBlock>(Incoming)) {
```
Code is gone in latest version

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -207,24 +213,69 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+                   "VPWidenPHIRecipe");
+            Incoming = R->getExiting();
+          }
+          if (Incoming != VPBB && !VPDT.dominates(VPBB, Incoming)) {
```
Code is gone in latest version

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-phi-accessors

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-phi-accessors

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-phi-accessors

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-phi-accessors

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-phi-accessors

fhahn committed:
!fixup address latest comments, thanks

fhahn committed:
Use virtual

[VPlan] Use VPPhiAccessors for VPIRPhi.

astOf

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-phi-accessors

fhahn committed:
!fixup address comments, make getAsRecipe pure virtual.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-verify-def-use-phi

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-verify-def-use-phi

fhahn committed:
!fixup update after merging f2e62cfca5e571.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-verify-def-use-phi

fhahn committed:
!fixup fix formatting

fhahn reviewed (commented):
Ping :)

Now that https://github.com/llvm/llvm-project/pull/138472 and https://github.com/llvm/llvm-project/pull/139151 landed, this should be ready for another look

david-arm commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1135,7 +1135,14 @@ class VPPhiAccessors {
 
   /// Returns the number of incoming values, also number of incoming blocks.
-  unsigned getNumIncoming() const { return getAsRecipe()->getNumOperands(); }
+  /// Note that at the moment, VPWidenIntOrFpInductionRecipes only have a single
```
Looks like this still includes bits of https://github.com/llvm/llvm-project/pull/138472, which looks like it hasn't landed yet?

fhahn replied:
#138472 simply adds additional iterators, but that is optional and not needed for the verification patch itself; they both need handling for additional recipes  in `getNumIncoming` though. Currently https://github.com/llvm/llvm-project/pull/138472 is more of a follow-up after this patch

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-verify-def-use-phi

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-verify-def-use-phi

fhahn committed:
!fxup improve verifier message, add verification test.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -207,24 +213,69 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+              continue;
+            const VPBlockBase *Incoming = UserVPBB->getPredecessors()[Idx];
+            if (Incoming != VPBB && !VPDT.dominates(VPBB, Incoming)) {
+              errs() << "Use before def!\n";
```
Updated to improve the message + test

ayalz commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,48 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
+        return cast<VPIRPhi>(R);
+      case VPDef::VPWidenPHISC:
+        return cast<VPWidenPHIRecipe>(R);
+      default:
+        return cast<VPHeaderPHIRecipe>(R);
```
An unreachable default, or one that returns null for !isPossible cases, trigger const_cast failure?

fhahn replied:
VPHeaderPhiRecipe includes multiple recipes, which is why I put it in the default, to avoid missing cases if new ones get added. The cast should already verify and assert that the cast is possible.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -220,12 +219,28 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+            const VPBasicBlock *IncVPBB = Phi->getIncomingBlock(Idx);
+            if (IncVPV != V)
+              continue;
+            if (IncVPBB != VPBB && !VPDT.dominates(VPBB, IncVPBB)) {
```
`dominates(VPBB, IncVPBB)` returns true when VPBB == IncVPBB, so suffice to check `if (!VPDT.dominates(VPBB, IncVPBB))`? But is it ok to have VPBB == IncVPBB, given that U is a phi - only if V is a phi too and such dependences between two phi's of the same block are allowed in VPlan, where fixed order dependences are modelled with intervening splices - for VF>1?
In any case good to continue with early continue:
```suggestion
            if (VPDT.properlyDominates(VPBB, IncVPBB))
              continue;
```

fhahn replied:
Updated to just `dominates`. We could have a loop with a single block, then the incoming value from the latch would be definined in the same block?

ayalz replied:
Ah, right; a value can be defined in the header and feed a header phi (of the same header block) even when the loop has additional blocks. Can check that in this case it's the 2nd incoming of the phi, i.e., across the backedge.

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -220,12 +219,28 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
-             cast<VPInstruction>(UI)->getOpcode() == Instruction::PHI))
+        if (auto *Phi = dyn_cast<VPPhiAccessors>(UI)) {
+          for (unsigned Idx = 0; Idx != Phi->getNumIncoming(); ++Idx) {
+            VPValue *IncVPV = Phi->getIncomingValue(Idx);
+            const VPBasicBlock *IncVPBB = Phi->getIncomingBlock(Idx);
+            if (IncVPV != V)
+              continue;
```
```suggestion
            VPValue *IncomingV = Phi->getIncomingValue(Idx);
            if (IncomingV != V)
              continue;
            const VPBasicBlock *IncomingVPBB = Phi->getIncomingBlock(Idx);
```

fhahn replied:
Updated, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -220,12 +219,28 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+          }
+          continue;
+        }
+        if (isa<VPPredInstPHIRecipe>(UI))
```
Anything to check?

fhahn replied:
Yep, added a TODO for now, thanks

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -207,24 +213,69 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+          continue;
+        }
 
         // If the user is in the same block, check it comes after R in the
```
independent typo.

fhahn replied:
Yep can fix separately, thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,48 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
 };
 
+/// Casting from VPRecipeBase -> VPPhiAccessors is supported for all recipe
+/// types implementing VPPhiAccessors.
```
Yep as there's no link between RecipeBase and VPPhiAccessors. `CastIsPossible`  is used for `isa`

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,48 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
+/// types implementing VPPhiAccessors.
+template <> struct CastIsPossible<VPPhiAccessors, const VPRecipeBase *> {
+  static inline bool isPossible(const VPRecipeBase *f) {
+    return isa<VPIRPhi, VPHeaderPHIRecipe, VPWidenPHIRecipe, VPPhi>(f);
```
Yes, but I left it out for now because more work will be needed as there the operands also don't match the # of predecessors.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,48 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
+/// Support casting from VPRecipeBase -> VPPhiAccessors, by down-casting to the
+/// recipe types implementing VPPhiAccessors.
+template <>
+struct CastInfo<VPPhiAccessors, const VPRecipeBase *>
```
This is used by `cast`/`dyn_cast` to perform the actual cast.

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-verify-def-use-phi

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,48 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
+
+  static inline VPPhiAccessors *castFailed() { return nullptr; }
+
+  static inline VPPhiAccessors *doCastIfPossible(const VPRecipeBase *f) {
```
This is used by `dyn_cast`

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,48 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
+
+  using Self = CastInfo<VPPhiAccessors, const VPRecipeBase *>;
+
+  using CastReturnType =
```
Unused in the latest version, removed, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -1135,7 +1135,14 @@ class VPPhiAccessors {
   /// Returns the number of incoming values, also number of incoming blocks.
-  unsigned getNumIncoming() const { return getAsRecipe()->getNumOperands(); }
+  /// Note that at the moment, VPWidenIntOrFpInductionRecipes only have a single
+  /// incoming value, its start value.
```
Done thanks

fhahn committed:
!fixup address latest comments, thanks!

ayalz commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -220,12 +219,29 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
-             cast<VPInstruction>(UI)->getOpcode() == Instruction::PHI))
+        if (auto *Phi = dyn_cast<VPPhiAccessors>(UI)) {
+          for (unsigned Idx = 0; Idx != Phi->getNumIncoming(); ++Idx) {
+            VPValue *IncVPV = Phi->getIncomingValue(Idx);
```
"Inc" is ambiguous - e.g., also stands for "Increment". Hence suggested renaming to `IncomingV` and `IncomingVPBB`.

fhahn replied:
updated, thanks

ayalz reviewed (approved):
LGTM, thanks!

fhahn committed:
Merge remote-tracking branch 'origin/main' into vplan-verify-def-use-phi

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,43 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
 };
 
+/// Casting from VPRecipeBase -> VPPhiAccessors is supported for all recipe
+/// types implementing VPPhiAccessors.
```
Adde a note, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,43 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
+/// Casting from VPRecipeBase -> VPPhiAccessors is supported for all recipe
+/// types implementing VPPhiAccessors.
+template <> struct CastIsPossible<VPPhiAccessors, const VPRecipeBase *> {
+  static inline bool isPossible(const VPRecipeBase *f) {
```
Added thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,43 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
+  }
+};
+/// Support casting from VPRecipeBase -> VPPhiAccessors, by down-casting to the
+/// recipe types implementing VPPhiAccessors.
```
Added thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -3283,6 +3295,43 @@ class VPScalarIVStepsRecipe : public VPRecipeWithIRFlags,
+    }());
+  }
+
+  static inline VPPhiAccessors *doCastIfPossible(const VPRecipeBase *f) {
```
added thanks

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:
```diff
@@ -220,12 +219,29 @@ bool VPlanVerifier::verifyVPBasicBlock(const VPBasicBlock *VPBB) {
+            if (IncVPV != V)
+              continue;
+            const VPBasicBlock *IncVPBB = Phi->getIncomingBlock(Idx);
+            if (!VPDT.dominates(VPBB, IncVPBB)) {
```
updated thanks

fhahn committed:
!fixup address latest comments, thanks!

vvereschaka commented:
@fhahn ,

these changes cause a crash for the following tests on the expensive check builders such as [llvm-clang-x86_64-expensive-checks-ubuntu](https://lab.llvm.org/buildbot/#/builders/187):
* first-order-recurrence.ll
* pointer_iv.ll
* pr48340.ll

more details here https://lab.llvm.org/buildbot/#/builders/187/builds/5956

```
/home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/build/bin/opt -passes=loop-vectorize --force-vector-width=4 --force-vector-interleave=0 -S -o - < /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/test/Transforms/LoopVectorize/X86/pr48340.ll | /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/build/bin/FileCheck /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/test/Transforms/LoopVectorize/X86/pr48340.ll # RUN: at line 2
+ /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/build/bin/opt -passes=loop-vectorize --force-vector-width=4 --force-vector-interleave=0 -S -o -
+ /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/build/bin/FileCheck /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/test/Transforms/LoopVectorize/X86/pr48340.ll
opt: /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlan.cpp:787: const llvm::VPBasicBlock* llvm::VPBasicBlock::getCFGPredecessor(unsigned int) const: Assertion `Idx < 2 && Region->getNumPredecessors() == 1 && "loop region has a single predecessor (preheader), its entry block " "has 2 incoming blocks"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/build/bin/opt -passes=loop-vectorize --force-vector-width=4 --force-vector-interleave=0 -S -o -
1.	Running pass "function(loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>)" on module "<stdin>"
2.	Running pass "loop-vectorize<no-interleave-forced-only;no-vectorize-forced-only;>" on function "foo"
 #0 0x00005581969b7064 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Support/Unix/Signals.inc:804:22
 #1 0x00005581969b74f7 PrintStackTraceSignalHandler(void*) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Support/Unix/Signals.inc:888:1
 #2 0x00005581969b48b3 llvm::sys::RunSignalHandlers() /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Support/Signals.cpp:105:20
 #3 0x00005581969b68d3 SignalHandler(int, siginfo_t*, void*) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Support/Unix/Signals.inc:418:13
 #4 0x00007fc006abf520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #5 0x00007fc006b139fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #6 0x00007fc006abf476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #7 0x00007fc006aa57f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #8 0x00007fc006aa571b (/lib/x86_64-linux-gnu/libc.so.6+0x2871b)
 #9 0x00007fc006ab6e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#10 0x0000558193c88f1b llvm::VPBasicBlock::getCFGPredecessor(unsigned int) const /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlan.cpp:793:21
#11 0x00005581939f1a4c llvm::VPPhiAccessors::getIncomingBlock(unsigned int) const /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlan.h:3487:1
#12 0x0000558193d3a979 (anonymous namespace)::VPlanVerifier::verifyVPBasicBlock(llvm::VPBasicBlock const*) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:228:69
#13 0x0000558193d3b32c (anonymous namespace)::VPlanVerifier::verifyBlock(llvm::VPBlockBase const*) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:362:16
#14 0x0000558193d3b452 (anonymous namespace)::VPlanVerifier::verifyBlocksInRegion(llvm::VPRegionBlock const*) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:373:9
#15 0x0000558193d3b5cf (anonymous namespace)::VPlanVerifier::verifyRegion(llvm::VPRegionBlock const*) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:393:37
#16 0x0000558193d3b670 (anonymous namespace)::VPlanVerifier::verifyRegionRec(llvm::VPRegionBlock const*) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:398:31
#17 0x0000558193d3b7e2 (anonymous namespace)::VPlanVerifier::verify(llvm::VPlan const&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:412:7
#18 0x0000558193d3baf2 llvm::verifyVPlanIsValid(llvm::VPlan const&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlanVerifier.cpp:461:25
#19 0x0000558193a0ae3d void llvm::VPlanTransforms::runPass<unsigned int, llvm::LLVMContext&>(void (*)(llvm::VPlan&, unsigned int, llvm::LLVMContext&), llvm::VPlan&, std::remove_reference<unsigned int>::type&, std::remove_reference<llvm::LLVMContext&>::type&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/VPlanTransforms.h:54:3
#20 0x00005581939c5c1f llvm::LoopVectorizationPlanner::executePlan(llvm::ElementCount, unsigned int, llvm::VPlan&, llvm::InnerLoopVectorizer&, llvm::DominatorTree*, bool) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7796:27
#21 0x00005581939d7f2c llvm::LoopVectorizePass::processLoop(llvm::Loop*) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:10980:24
#22 0x00005581939d859a llvm::LoopVectorizePass::runImpl(llvm::Function&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:11059:27
#23 0x00005581939d88e7 llvm::LoopVectorizePass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:11096:39
#24 0x0000558192f7e657 llvm::detail::PassModel<llvm::Function, llvm::LoopVectorizePass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/include/llvm/IR/PassManagerInternal.h:92:3
#25 0x00005581966d27ad llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/include/llvm/IR/PassManagerImpl.h:85:18
#26 0x0000558190a78fab llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/include/llvm/IR/PassManagerInternal.h:92:3
#27 0x00005581966d1682 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/lib/IR/PassManager.cpp:129:23
#28 0x0000558190a78d25 llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/include/llvm/IR/PassManagerInternal.h:92:3
#29 0x00005581966d23d9 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/include/llvm/IR/PassManagerImpl.h:85:18
#30 0x00005581901732f9 llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/tools/opt/NewPMDriver.cpp:557:10
#31 0x00005581901437eb optMain /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/tools/opt/optdriver.cpp:729:27
#32 0x0000558190141001 main /home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/llvm-project/llvm/tools/opt/opt.cpp:25:64
#33 0x00007fc006aa6d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#34 0x00007fc006aa6e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#35 0x0000558190140ee5 _start (/home/buildbot/worker/as-builder-4/ramdisk/expensive-checks/build/bin/opt+0xfdeee5)
FileCheck error: '<stdin>' is empty.
```

would you fix the problem as soon as it possible or revert these changes

fhahn commented:
@vvereschaka should be back to green after b3e7e4bb252c82c45639569c6b5f91c5c63742b6
